File: trunk/Yuki/Examples/Communication/CSharp/Src/Server/Servers/TcpServer.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using BaseSystem;
using Net;

namespace Server
{
    public partial class Streamed<TServerContext>
        where TServerContext : IServerContext
    {
        /// <summary>
        /// 本类的所有非继承的公共成员均是线程安全的。
        /// </summary>
        public class TcpServer : IServer
        {
            private class BindingInfo
            {
                public LockedVariable<Socket> Socket;
                public AsyncConsumer<SocketAsyncEventArgs> ListenConsumer;
                public Action Start;
            }

            private LockedVariable<Boolean> IsRunningValue = new LockedVariable<Boolean>(false);
            public Boolean IsRunning
            {
                get
                {
                    return IsRunningValue.Check(b => b);
                }
            }

            private class IpSessionInfo
            {
                public int Count = 0;
                public HashSet<TcpSession> Authenticated = new HashSet<TcpSession>();
            }

            private Dictionary<IPEndPoint, BindingInfo> BindingInfos = new Dictionary<IPEndPoint, BindingInfo>();
            private CancellationTokenSource ListeningTaskTokenSource;
            private AsyncConsumer<Socket> AcceptConsumer;
            private AsyncConsumer<TcpSession> PurifyConsumer;

            private class ServerSessionSets
            {
                public HashSet<TcpSession> Sessions = new HashSet<TcpSession>();
                public Dictionary<IPAddress, IpSessionInfo> IpSessions = new Dictionary<IPAddress, IpSessionInfo>();
            }
            private LockedVariable<ServerSessionSets> SessionSets = new LockedVariable<ServerSessionSets>(new ServerSessionSets());

            public TServerContext ServerContext { get; private set; }
            private Func<ISessionContext, IBinaryTransformer, KeyValuePair<IServerImplementation, IStreamedVirtualTransportServer>> VirtualTransportServerFactory;
            private Action<Action> QueueUserWorkItem;
            private Action<Action> PurifierQueueUserWorkItem;

            private int MaxBadCommandsValue = 8;
            private IPEndPoint[] BindingsValue = { };
            private int? SessionIdleTimeoutValue = null;
            private int? UnauthenticatedSessionIdleTimeoutValue = null;
            private int? MaxConnectionsValue = null;
            private int? MaxConnectionsPerIPValue = null;
            private int? MaxUnauthenticatedPerIPValue = null;

            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int MaxBadCommands
            {
                get
                {
                    return MaxBadCommandsValue;
                }
                set
                {
                    if (IsRunning) { throw new InvalidOperationException(); }
                    MaxBadCommandsValue = value;
                }
            }

            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public IPEndPoint[] Bindings
            {
                get
                {
                    return BindingsValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            BindingsValue = value;
                        }
                    );
                }
            }
            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int? SessionIdleTimeout
            {
                get
                {
                    return SessionIdleTimeoutValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            SessionIdleTimeoutValue = value;
                        }
                    );
                }
            }
            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int? UnauthenticatedSessionIdleTimeout
            {
                get
                {
                    return UnauthenticatedSessionIdleTimeoutValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            UnauthenticatedSessionIdleTimeoutValue = value;
                        }
                    );
                }
            }
            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int? MaxConnections
            {
                get
                {
                    return MaxConnectionsValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            MaxConnectionsValue = value;
                        }
                    );
                }
            }
            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int? MaxConnectionsPerIP
            {
                get
                {
                    return MaxConnectionsPerIPValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            MaxConnectionsPerIPValue = value;
                        }
                    );
                }
            }
            /// <summary>只能在启动前修改，以保证线程安全</summary>
            public int? MaxUnauthenticatedPerIP
            {
                get
                {
                    return MaxUnauthenticatedPerIPValue;
                }
                set
                {
                    IsRunningValue.DoAction
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }
                            MaxUnauthenticatedPerIPValue = value;
                        }
                    );
                }
            }

            public LockedVariable<Dictionary<ISessionContext, TcpSession>> SessionMappings = new LockedVariable<Dictionary<ISessionContext, TcpSession>>(new Dictionary<ISessionContext, TcpSession>());

            public TcpServer(TServerContext sc, Func<ISessionContext, IBinaryTransformer, KeyValuePair<IServerImplementation, IStreamedVirtualTransportServer>> VirtualTransportServerFactory, Action<Action> QueueUserWorkItem, Action<Action> PurifierQueueUserWorkItem)
            {
                ServerContext = sc;
                this.VirtualTransportServerFactory = VirtualTransportServerFactory;
                this.QueueUserWorkItem = QueueUserWorkItem;
                this.PurifierQueueUserWorkItem = PurifierQueueUserWorkItem;

                this.MaxConnectionsExceeded += OnMaxConnectionsExceeded;
                this.MaxConnectionsPerIPExceeded += OnMaxConnectionsPerIPExceeded;
            }

            private void OnMaxConnectionsExceeded(TcpSession s)
            {
                if (s != null && s.IsRunning)
                {
                    s.RaiseError("", "Client host rejected: too many connections, please try again later.");
                }
            }
            private void OnMaxConnectionsPerIPExceeded(TcpSession s)
            {
                if (s != null && s.IsRunning)
                {
                    s.RaiseError("", "Client host rejected: too many connections from your IP(" + s.RemoteEndPoint.Address + "), please try again later.");
                }
            }

            public void Start()
            {
                var Success = false;

                try
                {
                    IsRunningValue.Update
                    (
                        b =>
                        {
                            if (b) { throw new InvalidOperationException(); }

                            if (BindingsValue.Length == 0)
                            {
                                throw new Exception("NoValidBinding");
                            }

                            ListeningTaskTokenSource = new CancellationTokenSource();

                            var ListeningTaskToken = ListeningTaskTokenSource.Token;

                            Action<TcpSession> Purify = StoppingSession =>
                            {
                                SessionSets.DoAction
                                (
                                    ss =>
                                    {
                                        if (ss.Sessions.Contains(StoppingSession))
                                        {
                                            ss.Sessions.Remove(StoppingSession);
                                            var IpAddress = StoppingSession.RemoteEndPoint.Address;
                                            var isi = ss.IpSessions[IpAddress];
                                            if (isi.Authenticated.Contains(StoppingSession))
                                            {
                                                isi.Authenticated.Remove(StoppingSession);
                                            }
                                            isi.Count -= 1;
                                            if (isi.Count == 0)
                                            {
                                                ss.IpSessions.Remove(IpAddress);
                                            }
                                        }
                                    }
                                );
                                StoppingSession.Dispose();
                            };

                            Action<Socket> Accept = a =>
                            {
                                IPEndPoint e;
                                try
                                {
                                    e = (IPEndPoint)(a.RemoteEndPoint);
                                }
                                catch
                                {
                                    a.Dispose();
                                    return;
                                }
                                var s = new TcpSession(this, new StreamedAsyncSocket(a, UnauthenticatedSessionIdleTimeoutValue, QueueUserWorkItem), e, VirtualTransportServerFactory, QueueUserWorkItem);

                                if (MaxConnectionsValue.HasValue && (SessionSets.Check(ss => ss.Sessions.Count) >= MaxConnectionsValue.Value))
                                {
                                    PurifyConsumer.DoOne();
                                }
                                if (MaxConnectionsValue.HasValue && (SessionSets.Check(ss => ss.Sessions.Count) >= MaxConnectionsValue.Value))
                                {
                                    try
                                    {
                                        s.Start();
                                        if (MaxConnectionsExceeded != null)
                                        {
                                            MaxConnectionsExceeded(s);
                                        }
                                    }
                                    finally
                                    {
                                        s.Dispose();
                                    }
                                    return;
                                }

                                if (MaxConnectionsPerIPValue.HasValue && (SessionSets.Check(ss => ss.IpSessions.ContainsKey(e.Address) ? ss.IpSessions[e.Address].Count : 0) >= MaxConnectionsPerIPValue.Value))
                                {
                                    try
                                    {
                                        s.Start();
                                        if (MaxConnectionsPerIPExceeded != null)
                                        {
                                            MaxConnectionsPerIPExceeded(s);
                                        }
                                    }
                                    finally
                                    {
                                        PurifyConsumer.Push(s);
                                    }
                                    return;
                                }

                                if (MaxUnauthenticatedPerIPValue.HasValue && (SessionSets.Check(ss => ss.IpSessions.ContainsKey(e.Address) ? ss.IpSessions[e.Address].Count : 0) >= MaxUnauthenticatedPerIPValue.Value))
                                {
                                    try
                                    {
                                        s.Start();
                                        if (MaxConnectionsPerIPExceeded != null)
                                        {
                                            MaxConnectionsPerIPExceeded(s);
                                        }
                                    }
                                    finally
                                    {
                                        PurifyConsumer.Push(s);
                                    }
                                    return;
                                }

                                SessionSets.DoAction
                                (
                                    ss =>
                                    {
                                        ss.Sessions.Add(s);
                                        if (ss.IpSessions.ContainsKey(e.Address))
                                        {
                                            ss.IpSessions[e.Address].Count += 1;
                                        }
                                        else
                                        {
                                            var isi = new IpSessionInfo();
                                            isi.Count += 1;
                                            ss.IpSessions.Add(e.Address, isi);
                                        }
                                    }
                                );

                                s.Start();
                            };

                            var Exceptions = new List<Exception>();
                            foreach (var Binding in BindingsValue)
                            {
                                Func<Socket> CreateSocket = () =>
                                {
                                    var s = new Socket(Binding.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
                                    return s;
                                };

                                var Socket = CreateSocket();

                                try
                                {
                                    Socket.Bind(Binding);
                                }
                                catch (SocketException ex)
                                {
                                    Exceptions.Add(ex);
                                    continue;
                                }
                                Socket.Listen(MaxConnectionsValue.HasValue ? (MaxConnectionsValue.Value + 1) : 128);

                                var BindingInfo = new BindingInfo();
                                BindingInfo.Socket = new LockedVariable<Socket>(Socket);
                                Action Start = null;
                                Func<SocketAsyncEventArgs, Boolean> Completed = args =>
                                {
                                    try
                                    {
                                        if (ListeningTaskToken.IsCancellationRequested) { return false; }
                                        if (args.SocketError == SocketError.Success)
                                        {
                                            var a = args.AcceptSocket;
                                            QueueUserWorkItem(() => Accept(a));
                                        }
                                        else
                                        {
                                            BindingInfo.Socket.Update
                                            (
                                                OriginalSocket =>
                                                {
                                                    try
                                                    {
                                                        OriginalSocket.Dispose();
                                                    }
                                                    catch (Exception)
                                                    {
                                                    }
                                                    var NewSocket = CreateSocket();
                                                    NewSocket.Bind(Binding);
                                                    NewSocket.Listen(MaxConnectionsValue.HasValue ? (MaxConnectionsValue.Value + 1) : 128);
                                                    Socket = NewSocket;
                                                    return NewSocket;
                                                }
                                            );
                                        }
                                    }
                                    finally
                                    {
                                        args.Dispose();
                                    }
                                    Start();
                                    return true;
                                };
                                Start = () =>
                                {
                                    var EventArgs = new SocketAsyncEventArgs();
                                    var bs = BindingInfo.Socket.Check(s => s);
                                    EventArgs.Completed += (o, args) => BindingInfo.ListenConsumer.Push(args);
                                    try
                                    {
                                        if (!bs.AcceptAsync(EventArgs))
                                        {
                                            BindingInfo.ListenConsumer.Push(EventArgs);
                                        }
                                    }
                                    catch (ObjectDisposedException)
                                    {
                                    }
                                };
                                BindingInfo.ListenConsumer = new AsyncConsumer<SocketAsyncEventArgs>(QueueUserWorkItem, Completed, 1);
                                BindingInfo.Start = Start;

                                BindingInfos.Add(Binding, BindingInfo);
                            }
                            if (BindingInfos.Count == 0)
                            {
                                throw new AggregateException(Exceptions);
                            }

                            PurifyConsumer = new AsyncConsumer<TcpSession>(PurifierQueueUserWorkItem, s => { Purify(s); return true; }, int.MaxValue);

                            foreach (var BindingInfo in BindingInfos.Values)
                            {
                                BindingInfo.Start();
                            }

                            Success = true;

                            return true;
                        }
                    );
                }
                finally
                {
                    if (!Success)
                    {
                        Stop();
                    }
                }
            }

            public void Stop()
            {
                IsRunningValue.Update
                (
                    b =>
                    {
                        if (!b) { return false; }

                        if (ListeningTaskTokenSource != null)
                        {
                            ListeningTaskTokenSource.Cancel();
                        }
                        foreach (var BindingInfo in BindingInfos.Values)
                        {
                            BindingInfo.Socket.DoAction
                            (
                                Socket =>
                                {
                                    try
                                    {
                                        Socket.Dispose();
                                    }
                                    catch (Exception)
                                    {
                                    }
                                }
                            );
                        }
                        foreach (var BindingInfo in BindingInfos.Values)
                        {
                            BindingInfo.ListenConsumer.Dispose();
                        }
                        BindingInfos.Clear();
                        if (ListeningTaskTokenSource != null)
                        {
                            ListeningTaskTokenSource = null;
                        }

                        if (AcceptConsumer != null)
                        {
                            AcceptConsumer.Dispose();
                            AcceptConsumer = null;
                        }
                        SessionSets.DoAction
                        (
                            ss =>
                            {
                                foreach (var s in ss.Sessions)
                                {
                                    s.Dispose();
                                }
                                ss.Sessions.Clear();
                                ss.IpSessions.Clear();
                            }
                        );

                        if (PurifyConsumer != null)
                        {
                            PurifyConsumer.Dispose();
                            PurifyConsumer = null;
                        }

                        return false;
                    }
                );
            }

            public void NotifySessionQuit(TcpSession s)
            {
                PurifyConsumer.Push(s);
            }
            public void NotifySessionAuthenticated(TcpSession s)
            {
                var e = s.RemoteEndPoint;
                SessionSets.DoAction
                (
                    ss =>
                    {
                        if (ss.IpSessions.ContainsKey(e.Address))
                        {
                            var isi = ss.IpSessions[e.Address];
                            if (!isi.Authenticated.Contains(s))
                            {
                                isi.Authenticated.Add(s);
                            }
                        }
                    }
                );
            }

            private event Action<TcpSession> MaxConnectionsExceeded;
            private event Action<TcpSession> MaxConnectionsPerIPExceeded;

            public void Dispose()
            {
                Stop();
            }
        }
    }
}

File: trunk/Yuki/Examples/Communication/CPP/Src/Server/Servers/TcpServer.cpp
#include "TcpServer.h"
#include "TcpSession.h"

#include "BaseSystem/AutoRelease.h"
#include "BaseSystem/StringUtilities.h"

#include <limits>
#include <algorithm>
#include <chrono>

namespace Server
{
    TcpServer::TcpServer(asio::io_service &IoService, std::shared_ptr<IServerContext> sc, std::function<std::pair<std::shared_ptr<IServerImplementation>, std::shared_ptr<IStreamedVirtualTransportServer>>(std::shared_ptr<ISessionContext>, std::shared_ptr<IBinaryTransformer>)> VirtualTransportServerFactory, std::function<void(std::function<void()>)> QueueUserWorkItem, std::function<void(std::function<void()>)> PurifierQueueUserWorkItem)
        :
        IsRunningValue(false),
        IoService(IoService),
        SessionSets(std::make_shared<ServerSessionSets>()),
        VirtualTransportServerFactory(VirtualTransportServerFactory),
        QueueUserWorkItem(QueueUserWorkItem),
        PurifierQueueUserWorkItem(PurifierQueueUserWorkItem),
        MaxBadCommandsValue(8),
        SessionIdleTimeoutValue(Optional<int>::CreateNotHasValue()),
        UnauthenticatedSessionIdleTimeoutValue(Optional<int>::CreateNotHasValue()),
        MaxConnectionsValue(Optional<int>::CreateNotHasValue()),
        MaxConnectionsPerIPValue(Optional<int>::CreateNotHasValue()),
        MaxUnauthenticatedPerIPValue(Optional<int>::CreateNotHasValue()),
        TimeoutCheckPeriodValue(30),
        SessionMappings(std::make_shared<std::unordered_map<std::shared_ptr<ISessionContext>, std::shared_ptr<TcpSession>>>())
    {
        ServerContext(sc);
    }

    void TcpServer::OnMaxConnectionsExceeded(std::shared_ptr<TcpSession> s)
    {
        if (s != nullptr && s->IsRunning())
        {
            s->RaiseError(L"", L"Client host rejected: too many connections, please try again later.");
        }
    }
    void TcpServer::OnMaxConnectionsPerIPExceeded(std::shared_ptr<TcpSession> s)
    {
        if (s != nullptr && s->IsRunning())
        {
            s->RaiseError(L"", L"Client host rejected: too many connections from your IP(" + s2w(s->RemoteEndPoint.address().to_string()) + L"), please try again later.");
        }
    }

    void TcpServer::DoTimeoutCheck()
    {
        auto TimePeriod = std::chrono::seconds(std::max(TimeoutCheckPeriodValue, 1));
        auto WaitHandler = [this, TimePeriod](const asio::error_code& ec)
        {
            if (!ec)
            {
                if (UnauthenticatedSessionIdleTimeoutValue.HasValue)
                {
                    auto CheckTime = std::chrono::steady_clock::now() + std::chrono::seconds(-UnauthenticatedSessionIdleTimeoutValue.Value());
                    SessionSets.DoAction([&](std::shared_ptr<ServerSessionSets> ss)
                    {
                        for (auto s : ss->Sessions)
                        {
                            auto IpAddress = s->RemoteEndPoint.address();
                            auto isi = ss->IpSessions[IpAddress];
                            if (isi->Authenticated.count(s) == 0)
                            {
                                if (s->LastActiveTime() < CheckTime)
                                {
                                    PurifyConsumer->Push(s);
                                }
                            }
                        }
                    });
                }

                if (SessionIdleTimeoutValue.HasValue)
                {
                    auto CheckTime = std::chrono::steady_clock::now() + std::chrono::seconds(-SessionIdleTimeoutValue.Value());
                    SessionSets.DoAction([&](std::shared_ptr<ServerSessionSets> ss)
                    {
                        for (auto s : ss->Sessions)
                        {
                            auto IpAddress = s->RemoteEndPoint.address();
                            auto isi = ss->IpSessions[IpAddress];
                            if (isi->Authenticated.count(s) > 0)
                            {
                                if (s->LastActiveTime() < CheckTime)
                                {
                                    PurifyConsumer->Push(s);
                                }
                            }
                        }
                    });
                }

                DoTimeoutCheck();
            }
        };
        LastActiveTimeCheckTimer = std::make_shared<asio::steady_timer>(this->IoService);
        LastActiveTimeCheckTimer->expires_from_now(TimePeriod);
        LastActiveTimeCheckTimer->async_wait(WaitHandler);
    }

    void TcpServer::Start()
    {
        auto Success = false;
        BaseSystem::AutoRelease ar([&]()
        {
            if (!Success)
            {
                Stop();
            }
        });

        IsRunningValue.Update([&](bool b)
        {
            if (b) { throw std::logic_error("InvalidOperationException"); }

            if (BindingsValue->size() == 0)
            {
                throw std::logic_error("NoValidBinding");
            }

            ListeningTaskToken = std::make_shared<BaseSystem::CancellationToken>();

            auto Purify = [=](std::shared_ptr<TcpSession> StoppingSession)
            {
                SessionSets.DoAction([=](std::shared_ptr<ServerSessionSets> ss)
                {
                    if (ss->Sessions.count(StoppingSession) > 0)
                    {
                        ss->Sessions.erase(StoppingSession);
                        auto IpAddress = StoppingSession->RemoteEndPoint.address();
                        auto isi = ss->IpSessions[IpAddress];
                        if (isi->Authenticated.count(StoppingSession) > 0)
                        {
                            isi->Authenticated.erase(StoppingSession);
                        }
                        isi->Count -= 1;
                        if (isi->Count == 0)
                        {
                            ss->IpSessions.erase(IpAddress);
                        }
                    }
                });
                StoppingSession->Stop();
            };

            auto Accept = [=](std::shared_ptr<asio::ip::tcp::socket> a)
            {
                asio::error_code ec;
                auto ep = a->remote_endpoint(ec);
                if (ec)
                {
                    a->close(ec);
                    return;
                }
                auto s = std::make_shared<TcpSession>(*this, a, ep, VirtualTransportServerFactory, QueueUserWorkItem);

                if (MaxConnectionsValue.OnHasValue() && (SessionSets.Check<int>([=](std::shared_ptr<ServerSessionSets> ss) { return static_cast<int>(ss->Sessions.size()); }) >= MaxConnectionsValue.Value()))
                {
                    PurifyConsumer->DoOne();
                }
                if (MaxConnectionsValue.OnHasValue() && (SessionSets.Check<int>([=](std::shared_ptr<ServerSessionSets> ss) { return static_cast<int>(ss->Sessions.size()); }) >= MaxConnectionsValue.Value()))
                {
                    BaseSystem::AutoRelease ar([&]()
                    {
                        s = nullptr;
                    });
                    s->Start();
                    OnMaxConnectionsExceeded(s);
                    return;
                }

                if (MaxConnectionsPerIPValue.OnHasValue() && (SessionSets.Check<int>([=](std::shared_ptr<ServerSessionSets> ss) { return ss->IpSessions.count(ep.address()) > 0 ? ss->IpSessions[ep.address()]->Count : 0; }) >= MaxConnectionsPerIPValue.Value()))
                {
                    BaseSystem::AutoRelease ar([&]()
                    {
                        PurifyConsumer->Push(s);
                    });
                    s->Start();
                    OnMaxConnectionsPerIPExceeded(s);
                    return;
                }

                if (MaxUnauthenticatedPerIPValue.OnHasValue() && (SessionSets.Check<int>([=](std::shared_ptr<ServerSessionSets> ss) { return ss->IpSessions.count(ep.address()) > 0 ? ss->IpSessions[ep.address()]->Count : 0; }) >= MaxUnauthenticatedPerIPValue.Value()))
                {
                    BaseSystem::AutoRelease ar([&]()
                    {
                        PurifyConsumer->Push(s);
                    });
                    s->Start();
                    OnMaxConnectionsPerIPExceeded(s);
                    return;
                }

                SessionSets.DoAction([=](std::shared_ptr<ServerSessionSets> ss)
                {
                    ss->Sessions.insert(s);
                    if (ss->IpSessions.count(ep.address()) > 0)
                    {
                        ss->IpSessions[ep.address()]->Count += 1;
                    }
                    else
                    {
                        auto isi = std::make_shared<IpSessionInfo>();
                        isi->Count += 1;
                        ss->IpSessions[ep.address()] = isi;
                    }
                });

                s->Start();
            };
            AcceptConsumer = std::make_shared<BaseSystem::AsyncConsumer<std::shared_ptr<asio::ip::tcp::socket>>>(QueueUserWorkItem, [=](std::shared_ptr<asio::ip::tcp::socket> a) { Accept(a); return true; }, std::numeric_limits<int>::max());

            auto Exceptions = std::make_shared<std::vector<asio::error_code>>();
            for (auto Binding : *BindingsValue)
            {
                auto CreateSocket = [=]() -> std::shared_ptr<asio::ip::tcp::acceptor>
                {
                    auto s = std::make_shared<asio::ip::tcp::acceptor>(IoService, Binding);
                    return s;
                };

                auto Socket = CreateSocket();

                asio::error_code ec;
                Socket->listen(MaxConnectionsValue.OnHasValue() ? (MaxConnectionsValue.Value() + 1) : 128, ec);
                if (ec)
                {
                    Exceptions->push_back(ec);
                    continue;
                }

                auto BindingInfo = std::make_shared<class BindingInfo>();
                BindingInfo->Socket = std::make_shared<BaseSystem::LockedVariable<std::shared_ptr<asio::ip::tcp::acceptor>>>(Socket);
                auto BindingInfoPtr = &*BindingInfo;
                auto Completed = [this, Binding, CreateSocket, BindingInfoPtr](std::shared_ptr<AcceptResult> args) -> bool
                {
                    if (ListeningTaskToken->IsCancellationRequested()) { return false; }
                    if (!args->ec)
                    {
                        auto a = args->AcceptSocket;
                        AcceptConsumer->Push(a);
                    }
                    else
                    {
                        BindingInfoPtr->Socket->Update([=](std::shared_ptr<asio::ip::tcp::acceptor> OriginalSocket) -> std::shared_ptr<asio::ip::tcp::acceptor>
                        {
                            return nullptr;
                        });
                        BindingInfoPtr->Socket->Update([=](std::shared_ptr<asio::ip::tcp::acceptor> OriginalSocket) -> std::shared_ptr<asio::ip::tcp::acceptor>
                        {
                            auto NewSocket = CreateSocket();
                            NewSocket->listen(MaxConnectionsValue.OnHasValue() ? (MaxConnectionsValue.Value() + 1) : 128);
                            return NewSocket;
                        });
                    }
                    BindingInfoPtr->Start();
                    return true;
                };
                BindingInfo->ListenConsumer = std::make_shared<BaseSystem::AsyncConsumer<std::shared_ptr<AcceptResult>>>(QueueUserWorkItem, Completed, 1);
                BindingInfo->Start = [this, BindingInfoPtr]()
                {
                    auto a = std::make_shared<AcceptResult>();
                    a->AcceptSocket = std::make_shared<asio::ip::tcp::socket>(IoService);
                    auto RemoteEndPoint = std::make_shared<asio::ip::tcp::endpoint>();
                    auto bs = BindingInfoPtr->Socket->Check<std::shared_ptr<asio::ip::tcp::acceptor>>([](std::shared_ptr<asio::ip::tcp::acceptor> s) { return s; });
                    bs->async_accept(*a->AcceptSocket, *RemoteEndPoint, [this, a, RemoteEndPoint, BindingInfoPtr](const asio::error_code& ec)
                    {
                        if (ec == asio::error::operation_aborted) { return; }
                        a->ec = ec;
                        BindingInfoPtr->ListenConsumer->Push(a);
                    });
                };

                BindingInfos.push_back(BindingInfo);
            }
            if (BindingInfos.size() == 0)
            {
                throw std::logic_error("NoValidBinding");
            }

            PurifyConsumer = std::make_shared<BaseSystem::AsyncConsumer<std::shared_ptr<TcpSession>>>(PurifierQueueUserWorkItem, [=](std::shared_ptr<TcpSession> s) { Purify(s); return true; }, std::numeric_limits<int>::max());

            if (UnauthenticatedSessionIdleTimeoutValue.OnHasValue() || SessionIdleTimeoutValue.OnHasValue())
            {
                DoTimeoutCheck();
            }

            for (auto BindingInfo : BindingInfos)
            {
                BindingInfo->Start();
            }

            Success = true;

            return true;
        });
    }

    void TcpServer::Stop()
    {
        IsRunningValue.Update([&](bool b)
        {
            if (!b) { return false; }

            if (LastActiveTimeCheckTimer != nullptr)
            {
                LastActiveTimeCheckTimer = nullptr;
            }

            if (ListeningTaskToken != nullptr)
            {
                ListeningTaskToken->Cancel();
            }
            for (auto BindingInfo : BindingInfos)
            {
                BindingInfo->Socket->Update([=](std::shared_ptr<asio::ip::tcp::acceptor> OriginalSocket) -> std::shared_ptr<asio::ip::tcp::acceptor>
                {
                    OriginalSocket->close();
                    return nullptr;
                });
            }
            for (auto BindingInfo : BindingInfos)
            {
                BindingInfo->ListenConsumer = nullptr;
            }
            BindingInfos.clear();
            if (ListeningTaskToken != nullptr)
            {
                ListeningTaskToken = nullptr;
            }

            if (AcceptConsumer != nullptr)
            {
                AcceptConsumer = nullptr;
            }

            std::unordered_set<std::shared_ptr<TcpSession>> Sessions;
            SessionSets.DoAction([&](std::shared_ptr<ServerSessionSets> ss)
            {
                Sessions = ss->Sessions;
                ss->Sessions.clear();
                ss->IpSessions.clear();
            });
            for (auto s : Sessions)
            {
                s->Stop();
            }
            Sessions.clear();

            if (PurifyConsumer != nullptr)
            {
                PurifyConsumer = nullptr;
            }

            return false;
        });
    }

    void TcpServer::NotifySessionQuit(std::shared_ptr<TcpSession> s)
    {
        PurifyConsumer->Push(s);
    }
    void TcpServer::NotifySessionAuthenticated(std::shared_ptr<TcpSession> s)
    {
        auto e = s->RemoteEndPoint;
        SessionSets.DoAction([=](std::shared_ptr<ServerSessionSets> ss)
        {
            if (ss->IpSessions.count(e.address()) > 0)
            {
                auto isi = ss->IpSessions[e.address()];
                if (isi->Authenticated.count(s) == 0)
                {
                    isi->Authenticated.insert(s);
                }
            }
        });
    }

    TcpServer::~TcpServer()
    {
        Stop();
    }
}
